### ISSUE TITLE
How to follow multiple relations from a resource?

### ISSUE THREAD

**Issue by jarib** [basti1302/traverson#7]
Hi, and thanks for your work on this project! I'm a bit new to both node and hypermedia, so bear with me.

I'm trying out traverson for a HAL API I'm building, but can't figure out a good way of following multiple relations from a resource. Using the `orders` example, I may want to fetch both the `ea:basket` and `ea:customer` relations of the order I've just found.

These are the only ways to do it that I've found (code untested):

a. start from the root resource for every relation 

```javascript
api.newRequest().follow('ea:orders', 'ea:find')
   .withTemplateParameters({ id: 13 })
   .getResource(function(err, order) { ... });

api.newRequest().follow('ea:orders', 'ea:find', 'ea:basket')
   .withTemplateParameters({ id: 13 })
   .getResource(function(err, basket) { ... });

api.newRequest().follow('ea:orders', 'ea:find', 'ea:customer')
   .withTemplateParameters({ id: 13 })
   .getResource(function(err, customer) { ... });
```

b. create a new traverson instance based on the `order` response

```javascript
api.newRequest().follow('ea:orders', 'ea:find')
   .withTemplateParameters({ id: 13 })
   .getResource(function(err, order) { 
     var orderApi = traverson.jsonHal.from(order._links.self.href);
     orderApi.newRequest().follow('ea:basket').getResource(function (err, basket) { ... });
     orderApi.newRequest().follow('ea:customer').getResource(function (err, customer) { ... });
   });
```

Here's a few ideas on how these could be made simpler:

c. re-use request with additional `.follow` calls (actual requests equivalent to ex. a)

```javascript
var order = api.newRequest()
            .follow('ea:orders', 'ea:find')
            .withTemplateParameters({ id: 13 })
            
            
order.getResource(...)
order.follow('ea:basket').getResource(...)
order.follow('ea:customer').getResource(...)
```

d. have the resource object double as a client (actual requests equivalent to ex. b)

```javascript
api.newRequest()
  .follow('ea:orders', 'ea:find')
  .withTemplateParameters({ id: 13 })
  .getResource(function(err, order) { 
     order.follow('ea:basket').getResource(...);
     order.follow('ea:customer').getResource(...)
  });
```

All of these feel suboptimal in different ways, both when it comes to performance (number of requests) and readability. It also seems like I'd have to do quite a bit of trickery to get a composed response object (consisting of order, basket and customer) that could be passed to a view or presenter. Though that may also just be my lack of node experience.

Going off into total fantasy for a moment, what I'm really after is a way of specifying all the relations I want up front, and simply get a composed response back. It'd have to be a quiet generic specification to support all the cases, but e.g. something like this:

```javascript
var requestSpec = [
  { 
    rel: "ea:orders",
    follow: [
      {
        rel: "ea:find",
        params: {id: 13},
        follow: [{rel: "ea:basket"}, {rel: "ea:customer"}]
      }
    ]
  }
]

api.request(requestSpec).get(function(err, resp) {
  resp['ea:orders']                              // result of /orders
  resp['ea:orders']['ea:find']                    // result of /orders/13 wrapped, wrapped in array (since you may want to follow ea:find multiple times)
  resp['ea:orders']['ea:find'][0]['ea:basket']    // result of /baskets/321 wrapped in array
  resp['ea:orders']['ea:find'][0]['ea:customer']  // result of /customers/42 wrapped in array
});
```

Or possibly sugar it a bit

```javascript
var order = api.follow("ea:orders").follow("ea:find", {id: 13})

order.follow('ea:basket')
order.follow('ea:customer')

order.getResources(...);
```

Perhaps something like this could be built on top of traverson. Any thoughts?

**Comment by jinder** [comment]
+1 for this.  It'd be nice to have a continuation option from the promise itself.  For example:

   .follow("datasources").getResource().then((result, traverson) =>  {
      return traverson.follow("test").getResource();
    }).then(......);

**Comment by basti1302** [comment]
I already started to work continuing traversals in [this branch](https://github.com/basti1302/traverson/tree/continue-traversal) (better late than never, right? ;-) ). I'd love some feedback on the API (which is work in progress right now), so chime in to the discussion at issue #40 if you like.

**Comment by basti1302** [comment]
Continuing link traversals has landed in 2.0.0, which should also enable the use case described in the original ticket.
